// Data.String.Regex/index.js

/**
 *  | Wraps Javascript's `RegExp` object that enables matching strings with
 *  | patternes defined by regular expressions.
 *  | For details of the underlying implementation, see [RegExp Reference at MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp).
 */
"use strict";
var $foreign = require("./foreign");
var Prelude = require("Prelude");
var Data_String = require("Data.String");
var Data_Maybe = require("Data.Maybe");
var showRegex = new Prelude.Show($foreign["showRegex'"]);

/**
 *  | Returns the string representation of the given `RegexFlags`.
 */
var renderFlags = function (f) {
    return (function () {
        if (f.global) {
            return "g";
        };
        if (!f.global) {
            return "";
        };
        throw new Error("Failed pattern match at bower_components/purescript-strings/src/Data/String/Regex.purs line 63, column 1 - line 64, column 1: " + [ f.global.constructor.name ]);
    })() + ((function () {
        if (f.ignoreCase) {
            return "i";
        };
        if (!f.ignoreCase) {
            return "";
        };
        throw new Error("Failed pattern match at bower_components/purescript-strings/src/Data/String/Regex.purs line 63, column 1 - line 64, column 1: " + [ f.ignoreCase.constructor.name ]);
    })() + ((function () {
        if (f.multiline) {
            return "m";
        };
        if (!f.multiline) {
            return "";
        };
        throw new Error("Failed pattern match at bower_components/purescript-strings/src/Data/String/Regex.purs line 63, column 1 - line 64, column 1: " + [ f.multiline.constructor.name ]);
    })() + ((function () {
        if (f.sticky) {
            return "y";
        };
        if (!f.sticky) {
            return "";
        };
        throw new Error("Failed pattern match at bower_components/purescript-strings/src/Data/String/Regex.purs line 63, column 1 - line 64, column 1: " + [ f.sticky.constructor.name ]);
    })() + (function () {
        if (f.unicode) {
            return "u";
        };
        if (!f.unicode) {
            return "";
        };
        throw new Error("Failed pattern match at bower_components/purescript-strings/src/Data/String/Regex.purs line 63, column 1 - line 64, column 1: " + [ f.unicode.constructor.name ]);
    })())));
};

/**
 *  | Constructs a `Regex` from a pattern string and flags.
 */
var regex = function (s) {
    return function (f) {
        return $foreign["regex'"](s)(renderFlags(f));
    };
};

/**
 *  | Parses the string representation of `RegexFlags`.
 */
var parseFlags = function (s) {
    return {
        global: Data_String.contains("g")(s), 
        ignoreCase: Data_String.contains("i")(s), 
        multiline: Data_String.contains("m")(s), 
        sticky: Data_String.contains("y")(s), 
        unicode: Data_String.contains("u")(s)
    };
};

/**
 *  | All flags set to false.
 */
var noFlags = {
    global: false, 
    ignoreCase: false, 
    multiline: false, 
    sticky: false, 
    unicode: false
};

/**
 *  | Matches the string against the `Regex` and returns an array of matches
 *  | if there were any. Each match has type `Maybe String`, where `Nothing`
 *  | represents an unmatched optional capturing group.
 *  | See [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match).
 */
var match = $foreign._match(Data_Maybe.Just.create)(Data_Maybe.Nothing.value);
module.exports = {
    noFlags: noFlags, 
    match: match, 
    parseFlags: parseFlags, 
    renderFlags: renderFlags, 
    regex: regex, 
    showRegex: showRegex, 
    split: $foreign.split, 
    search: $foreign.search, 
    "replace'": $foreign["replace'"], 
    replace: $foreign.replace, 
    test: $foreign.test, 
    flags: $foreign.flags, 
    source: $foreign.source
};
