// Test.Spec.Reporter/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Eff_Exception = require("Control.Monad.Eff.Exception");
var Data_Foldable = require("Data.Foldable");
var Data_Array = require("Data.Array");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Test_Spec = require("Test.Spec");
var Data_Monoid = require("Data.Monoid");
var Describe = (function () {
    function Describe(value0) {
        this.value0 = value0;
    };
    Describe.create = function (value0) {
        return new Describe(value0);
    };
    return Describe;
})();
var It = (function () {
    function It(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    It.create = function (value0) {
        return function (value1) {
            return new It(value0, value1);
        };
    };
    return It;
})();
var Pending = (function () {
    function Pending(value0) {
        this.value0 = value0;
    };
    Pending.create = function (value0) {
        return new Pending(value0);
    };
    return Pending;
})();
var showEntry = new Prelude.Show(function (_152) {
    if (_152 instanceof Describe) {
        return "Describe \"" + (Data_Foldable.intercalate(Data_Foldable.foldableArray)(Data_Monoid.monoidString)(" » ")(_152.value0) + "\"");
    };
    if (_152 instanceof It && _152.value1 instanceof Test_Spec.Success) {
        return "It \"" + (_152.value0 + "\" Success");
    };
    if (_152 instanceof It && _152.value1 instanceof Test_Spec.Failure) {
        return "It \"" + (_152.value0 + ("\" (Failure \"" + (Control_Monad_Eff_Exception.message(_152.value1.value0) + "\")")));
    };
    if (_152 instanceof Pending) {
        return "Pending \"" + (_152.value0 + "\"");
    };
    throw new Error("Failed pattern match at bower_components/purescript-spec/src/Test/Spec/Reporter.purs line 28, column 1 - line 34, column 1: " + [ _152.constructor.name ]);
});
var eqEntry = new Prelude.Eq(function (_150) {
    return function (_151) {
        if (_150 instanceof Describe && _151 instanceof Describe) {
            return Prelude["=="](Prelude.eqArray(Prelude.eqString))(_150.value0)(_151.value0);
        };
        if (_150 instanceof It && (_150.value1 instanceof Test_Spec.Success && (_151 instanceof It && _151.value1 instanceof Test_Spec.Success))) {
            return Prelude["=="](Prelude.eqString)(_150.value0)(_151.value0);
        };
        if (_150 instanceof It && (_150.value1 instanceof Test_Spec.Failure && (_151 instanceof It && _151.value1 instanceof Test_Spec.Failure))) {
            return Prelude["&&"](Prelude.booleanAlgebraBoolean)(Prelude["=="](Prelude.eqString)(_150.value0)(_151.value0))(Prelude["=="](Prelude.eqString)(Control_Monad_Eff_Exception.message(_150.value1.value0))(Control_Monad_Eff_Exception.message(_151.value1.value0)));
        };
        if (_150 instanceof It && _151 instanceof It) {
            return false;
        };
        if (_150 instanceof Pending && _151 instanceof Pending) {
            return Prelude["=="](Prelude.eqString)(_150.value0)(_151.value0);
        };
        throw new Error("Failed pattern match at bower_components/purescript-spec/src/Test/Spec/Reporter.purs line 20, column 1 - line 28, column 1: " + [ _150.constructor.name, _151.constructor.name ]);
    };
});
var countDescribes = function (groups) {
    var f = function (c) {
        return function (_153) {
            if (_153 instanceof Describe) {
                return Prelude["+"](Prelude.semiringInt)(c)(1);
            };
            return c;
        };
    };
    return Data_Foldable.foldl(Data_Foldable.foldableArray)(f)(0)(groups);
};
var collapse = function (_149) {
    if (_149 instanceof Test_Spec.It) {
        return [ new It(_149.value0, _149.value1) ];
    };
    if (_149 instanceof Test_Spec.Pending) {
        return [ new Pending(_149.value0) ];
    };
    if (_149 instanceof Test_Spec.Describe) {
        var sub = Data_Array.concatMap(collapse)(_149.value1);
        var prependName = function (_154) {
            if (_154 instanceof Describe) {
                return Describe.create(Data_Array.cons(_149.value0)(_154.value0));
            };
            return _154;
        };
        var c = countDescribes(sub);
        var _269 = Prelude["=="](Prelude.eqInt)(c)(0);
        if (_269) {
            return Data_Array.cons(new Describe([ _149.value0 ]))(sub);
        };
        if (!_269) {
            return Prelude.map(Prelude.functorArray)(prependName)(sub);
        };
        throw new Error("Failed pattern match: " + [ _269.constructor.name ]);
    };
    throw new Error("Failed pattern match: " + [ _149.constructor.name ]);
};
module.exports = {
    Describe: Describe, 
    It: It, 
    Pending: Pending, 
    collapse: collapse, 
    eqEntry: eqEntry, 
    showEntry: showEntry
};
