// Test.QuickCheck.Arbitrary/index.js
"use strict";
var Prelude = require("Prelude");
var Test_QuickCheck_Gen = require("Test.QuickCheck.Gen");
var Data_Int = require("Data.Int");
var Data_String = require("Data.String");
var Data_Char = require("Data.Char");
var Data_Foldable = require("Data.Foldable");
var Data_Either = require("Data.Either");
var Data_Maybe = require("Data.Maybe");
var Data_Tuple = require("Data.Tuple");

/**
 *  | The `Arbitrary` class represents those types whose values can be
 *  | _randomly-generated_.
 *  |
 *  | `arbitrary` uses the `Gen` monad to express a random generator for
 *  | the type `t`. Combinators in the `Test.QuickCheck.Gen`
 *  | module can be used to construct random generators.
 */
var Arbitrary = function (arbitrary) {
    this.arbitrary = arbitrary;
};

/**
 *  | The `Coarbitrary` class represents types which appear on the left of
 *  | an `Arbitrary` function arrow.
 *  |
 *  | To construct an `Arbitrary` instance for the type `a -> b`, we need to
 *  | use the input of type `a` to _perturb_ a random generator for `b`. This
 *  | is the role of the `coarbitrary` function.
 *  |
 *  | `Coarbitrary` instances can be written using the `perturbGen` function.
 */
var Coarbitrary = function (coarbitrary) {
    this.coarbitrary = coarbitrary;
};

/**
 *  | The `Coarbitrary` class represents types which appear on the left of
 *  | an `Arbitrary` function arrow.
 *  |
 *  | To construct an `Arbitrary` instance for the type `a -> b`, we need to
 *  | use the input of type `a` to _perturb_ a random generator for `b`. This
 *  | is the role of the `coarbitrary` function.
 *  |
 *  | `Coarbitrary` instances can be written using the `perturbGen` function.
 */
var coarbitrary = function (dict) {
    return dict.coarbitrary;
};
var coarbUnit = new Coarbitrary(function (_582) {
    return Test_QuickCheck_Gen.perturbGen(1.0);
});
var coarbTuple = function (__dict_Coarbitrary_0) {
    return function (__dict_Coarbitrary_1) {
        return new Coarbitrary(function (_584) {
            return Prelude[">>>"](Prelude.semigroupoidFn)(coarbitrary(__dict_Coarbitrary_0)(_584.value0))(coarbitrary(__dict_Coarbitrary_1)(_584.value1));
        });
    };
};
var coarbOrdering = new Coarbitrary(function (_583) {
    if (_583 instanceof Prelude.LT) {
        return Test_QuickCheck_Gen.perturbGen(1.0);
    };
    if (_583 instanceof Prelude.EQ) {
        return Test_QuickCheck_Gen.perturbGen(2.0);
    };
    if (_583 instanceof Prelude.GT) {
        return Test_QuickCheck_Gen.perturbGen(3.0);
    };
    throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 76, column 1 - line 81, column 1: " + [ _583.constructor.name ]);
});
var coarbNumber = new Coarbitrary(Test_QuickCheck_Gen.perturbGen);
var coarbMaybe = function (__dict_Coarbitrary_2) {
    return new Coarbitrary(function (_585) {
        if (_585 instanceof Data_Maybe.Nothing) {
            return Test_QuickCheck_Gen.perturbGen(1.0);
        };
        if (_585 instanceof Data_Maybe.Just) {
            return coarbitrary(__dict_Coarbitrary_2)(_585.value0);
        };
        throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 106, column 1 - line 110, column 1: " + [ _585.constructor.name ]);
    });
};
var coarbInt = new Coarbitrary(Prelude["<<<"](Prelude.semigroupoidFn)(Test_QuickCheck_Gen.perturbGen)(Data_Int.toNumber));
var coarbEither = function (__dict_Coarbitrary_3) {
    return function (__dict_Coarbitrary_4) {
        return new Coarbitrary(function (_586) {
            if (_586 instanceof Data_Either.Left) {
                return coarbitrary(__dict_Coarbitrary_3)(_586.value0);
            };
            if (_586 instanceof Data_Either.Right) {
                return coarbitrary(__dict_Coarbitrary_4)(_586.value0);
            };
            throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 115, column 1 - line 117, column 39: " + [ _586.constructor.name ]);
        });
    };
};
var coarbChar = new Coarbitrary(function (c) {
    return coarbitrary(coarbInt)(Data_Char.toCharCode(c));
});
var coarbBoolean = new Coarbitrary(function (_581) {
    if (_581) {
        return Test_QuickCheck_Gen.perturbGen(1.0);
    };
    if (!_581) {
        return Test_QuickCheck_Gen.perturbGen(2.0);
    };
    throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 39, column 1 - line 43, column 1: " + [ _581.constructor.name ]);
});
var coarbArray = function (__dict_Coarbitrary_5) {
    return new Coarbitrary(Data_Foldable.foldl(Data_Foldable.foldableArray)(function (f) {
        return function (x) {
            return Prelude["<<<"](Prelude.semigroupoidFn)(f)(coarbitrary(__dict_Coarbitrary_5)(x));
        };
    })(Prelude.id(Prelude.categoryFn)));
};
var coarbString = new Coarbitrary(function (s) {
    return coarbitrary(coarbArray(coarbMaybe(coarbInt)))(Prelude["<$>"](Prelude.functorArray)(Data_String.charCodeAt(Prelude.zero(Prelude.semiringInt)))(Data_String.split("")(s)));
});

/**
 *  | The `Arbitrary` class represents those types whose values can be
 *  | _randomly-generated_.
 *  |
 *  | `arbitrary` uses the `Gen` monad to express a random generator for
 *  | the type `t`. Combinators in the `Test.QuickCheck.Gen`
 *  | module can be used to construct random generators.
 */
var arbitrary = function (dict) {
    return dict.arbitrary;
};
var arbUnit = new Arbitrary(Prelude["return"](Test_QuickCheck_Gen.applicativeGen)(Prelude.unit));
var arbTuple = function (__dict_Arbitrary_8) {
    return function (__dict_Arbitrary_9) {
        return new Arbitrary(Prelude["<*>"](Test_QuickCheck_Gen.applyGen)(Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_Tuple.Tuple.create)(arbitrary(__dict_Arbitrary_8)))(arbitrary(__dict_Arbitrary_9)));
    };
};
var arbOrdering = new Arbitrary(Test_QuickCheck_Gen.oneOf(Prelude.pure(Test_QuickCheck_Gen.applicativeGen)(Prelude.LT.value))([ Prelude.pure(Test_QuickCheck_Gen.applicativeGen)(Prelude.EQ.value), Prelude.pure(Test_QuickCheck_Gen.applicativeGen)(Prelude.GT.value) ]));
var arbNumber = new Arbitrary(Test_QuickCheck_Gen.uniform);
var arbInt = new Arbitrary(Test_QuickCheck_Gen.chooseInt(Prelude.negate(Prelude.ringInt)(1000000))(1000000));
var arbFunction = function (__dict_Coarbitrary_11) {
    return function (__dict_Arbitrary_12) {
        return new Arbitrary(Test_QuickCheck_Gen.repeatable(function (a) {
            return coarbitrary(__dict_Coarbitrary_11)(a)(arbitrary(__dict_Arbitrary_12));
        }));
    };
};
var arbChar = new Arbitrary(Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_Char.fromCharCode)(Test_QuickCheck_Gen.chooseInt(0)(65536)));
var arbBoolean = new Arbitrary(Prelude.bind(Test_QuickCheck_Gen.bindGen)(Test_QuickCheck_Gen.uniform)(function (_83) {
    return Prelude["return"](Test_QuickCheck_Gen.applicativeGen)(_83 * 2.0 < 1.0);
}));
var arbEither = function (__dict_Arbitrary_13) {
    return function (__dict_Arbitrary_14) {
        return new Arbitrary(Prelude.bind(Test_QuickCheck_Gen.bindGen)(arbitrary(arbBoolean))(function (_86) {
            if (_86) {
                return Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_Either.Left.create)(arbitrary(__dict_Arbitrary_13));
            };
            if (!_86) {
                return Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_Either.Right.create)(arbitrary(__dict_Arbitrary_14));
            };
            throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 110, column 1 - line 115, column 1: " + [ _86.constructor.name ]);
        }));
    };
};
var arbMaybe = function (__dict_Arbitrary_10) {
    return new Arbitrary(Prelude.bind(Test_QuickCheck_Gen.bindGen)(arbitrary(arbBoolean))(function (_85) {
        if (_85) {
            return Prelude.pure(Test_QuickCheck_Gen.applicativeGen)(Data_Maybe.Nothing.value);
        };
        if (!_85) {
            return Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_Maybe.Just.create)(arbitrary(__dict_Arbitrary_10));
        };
        throw new Error("Failed pattern match at bower_components/purescript-quickcheck/src/Test/QuickCheck/Arbitrary.purs line 101, column 1 - line 106, column 1: " + [ _85.constructor.name ]);
    }));
};
var arbArray = function (__dict_Arbitrary_15) {
    return new Arbitrary(Test_QuickCheck_Gen.arrayOf(arbitrary(__dict_Arbitrary_15)));
};
var arbString = new Arbitrary(Prelude["<$>"](Test_QuickCheck_Gen.functorGen)(Data_String.fromCharArray)(arbitrary(arbArray(arbChar))));
var coarbFunction = function (__dict_Arbitrary_6) {
    return function (__dict_Coarbitrary_7) {
        return new Coarbitrary(function (f) {
            return function (gen) {
                return Prelude.bind(Test_QuickCheck_Gen.bindGen)(arbitrary(arbArray(__dict_Arbitrary_6)))(function (_84) {
                    return coarbitrary(coarbArray(__dict_Coarbitrary_7))(Prelude.map(Prelude.functorArray)(f)(_84))(gen);
                });
            };
        });
    };
};
module.exports = {
    Coarbitrary: Coarbitrary, 
    Arbitrary: Arbitrary, 
    coarbitrary: coarbitrary, 
    arbitrary: arbitrary, 
    arbBoolean: arbBoolean, 
    coarbBoolean: coarbBoolean, 
    arbNumber: arbNumber, 
    coarbNumber: coarbNumber, 
    arbInt: arbInt, 
    coarbInt: coarbInt, 
    arbString: arbString, 
    coarbString: coarbString, 
    arbChar: arbChar, 
    coarbChar: coarbChar, 
    arbUnit: arbUnit, 
    coarbUnit: coarbUnit, 
    arbOrdering: arbOrdering, 
    coarbOrdering: coarbOrdering, 
    arbArray: arbArray, 
    coarbArray: coarbArray, 
    arbFunction: arbFunction, 
    coarbFunction: coarbFunction, 
    arbTuple: arbTuple, 
    coarbTuple: coarbTuple, 
    arbMaybe: arbMaybe, 
    coarbMaybe: coarbMaybe, 
    arbEither: arbEither, 
    coarbEither: coarbEither
};
