// Control.Monad.List.Trans/index.js

/**
 *  | This module defines the list monad transformer, `ListT`.
 */
"use strict";
var Prelude = require("Prelude");
var Data_Lazy = require("Data.Lazy");
var Data_Maybe = require("Data.Maybe");
var Data_Tuple = require("Data.Tuple");
var Control_Monad_Trans = require("Control.Monad.Trans");
var Control_Monad_Eff_Class = require("Control.Monad.Eff.Class");
var Control_Alt = require("Control.Alt");
var Control_Alternative = require("Control.Alternative");
var Control_MonadPlus = require("Control.MonadPlus");
var Control_Plus = require("Control.Plus");
var Data_Monoid = require("Data.Monoid");
var Data_Unfoldable = require("Data.Unfoldable");

/**
 *  | The result of a single step in a `ListT` computation. Either:
 *  |
 *  | - Computation has finished (`Done`), or
 *  | - A result has been returned, along with the next part of the computation (`Yield`).
 *  |
 *  | The `Skip` constructor allows us to avoid traversing lists during certain operations.
 */
var Yield = (function () {
    function Yield(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Yield.create = function (value0) {
        return function (value1) {
            return new Yield(value0, value1);
        };
    };
    return Yield;
})();

/**
 *  | The result of a single step in a `ListT` computation. Either:
 *  |
 *  | - Computation has finished (`Done`), or
 *  | - A result has been returned, along with the next part of the computation (`Yield`).
 *  |
 *  | The `Skip` constructor allows us to avoid traversing lists during certain operations.
 */
var Skip = (function () {
    function Skip(value0) {
        this.value0 = value0;
    };
    Skip.create = function (value0) {
        return new Skip(value0);
    };
    return Skip;
})();

/**
 *  | The result of a single step in a `ListT` computation. Either:
 *  |
 *  | - Computation has finished (`Done`), or
 *  | - A result has been returned, along with the next part of the computation (`Yield`).
 *  |
 *  | The `Skip` constructor allows us to avoid traversing lists during certain operations.
 */
var Done = (function () {
    function Done() {

    };
    Done.value = new Done();
    return Done;
})();

/**
 *  | The list monad transformer.
 *  |
 *  | This monad transformer extends the base monad with _non-determinism_.
 *  | That is, the transformed monad supports the same effects as the base monad
 *  | but with multiple return values.
 */
var ListT = (function () {
    function ListT(value0) {
        this.value0 = value0;
    };
    ListT.create = function (value0) {
        return new ListT(value0);
    };
    return ListT;
})();

/**
 *  | Defer evaluation of a list.
 */
var wrapLazy = function (__dict_Applicative_0) {
    return function (v) {
        return ListT.create(Prelude.pure(__dict_Applicative_0)(new Skip(v)));
    };
};

/**
 *  | Lift a computation from the base monad.
 */
var wrapEffect = function (__dict_Functor_1) {
    return function (v) {
        return ListT.create(Prelude["<$>"](__dict_Functor_1)(Prelude["<<<"](Prelude.semigroupoidFn)(Skip.create)(Prelude["<<<"](Prelude.semigroupoidFn)(Data_Lazy.defer)(Prelude["const"])))(v));
    };
};
var unfold = function (__dict_Monad_2) {
    return function (f) {
        return function (z) {
            var g = function (_375) {
                if (_375 instanceof Data_Maybe.Just) {
                    return new Yield(_375.value0.value1, Data_Lazy.defer(function (_368) {
                        return unfold(__dict_Monad_2)(f)(_375.value0.value0);
                    }));
                };
                if (_375 instanceof Data_Maybe.Nothing) {
                    return Done.value;
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 115, column 3 - line 116, column 3: " + [ _375.constructor.name ]);
            };
            return ListT.create(Prelude["<$>"](((__dict_Monad_2["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(g)(f(z)));
        };
    };
};

/**
 *  | Run a computation in the `ListT` monad.
 */
var runListT = function (_371) {
    return _371.value0;
};

/**
 *  | Fold a list from the left, accumulating the list of results using the specified function.
 */
var scanl = function (__dict_Monad_3) {
    return function (f) {
        return function (b) {
            return function (l) {
                var g = function (_385) {
                    var h = function (_386) {
                        if (_386 instanceof Yield) {
                            var b$prime = f(_385.value0)(_386.value0);
                            return Data_Maybe.Just.create(new Data_Tuple.Tuple(new Data_Tuple.Tuple(b$prime, Data_Lazy.force(_386.value1)), b$prime));
                        };
                        if (_386 instanceof Skip) {
                            return Data_Maybe.Just.create(new Data_Tuple.Tuple(new Data_Tuple.Tuple(_385.value0, Data_Lazy.force(_386.value0)), _385.value0));
                        };
                        if (_386 instanceof Done) {
                            return Data_Maybe.Nothing.value;
                        };
                        throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 214, column 5 - line 216, column 5: " + [ _386.constructor.name ]);
                    };
                    return Prelude["<$>"](((__dict_Monad_3["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(h)(runListT(_385.value1));
                };
                return unfold(__dict_Monad_3)(g)(new Data_Tuple.Tuple(b, l));
            };
        };
    };
};

/**
 *  | Lift a computation on list steps to a computation on whole lists.
 */
var stepMap = function (__dict_Functor_4) {
    return function (f) {
        return function (l) {
            return ListT.create(Prelude["<$>"](__dict_Functor_4)(f)(runListT(l)));
        };
    };
};
var takeWhile = function (__dict_Applicative_5) {
    return function (f) {
        var g = function (_377) {
            if (_377 instanceof Yield) {
                var _681 = f(_377.value0);
                if (_681) {
                    return new Yield(_377.value0, Prelude["<$>"](Data_Lazy.functorLazy)(takeWhile(__dict_Applicative_5)(f))(_377.value1));
                };
                if (!_681) {
                    return Done.value;
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 139, column 3 - line 140, column 3: " + [ _681.constructor.name ]);
            };
            if (_377 instanceof Skip) {
                return Skip.create(Prelude["<$>"](Data_Lazy.functorLazy)(takeWhile(__dict_Applicative_5)(f))(_377.value0));
            };
            if (_377 instanceof Done) {
                return Done.value;
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 139, column 3 - line 140, column 3: " + [ _377.constructor.name ]);
        };
        return stepMap((__dict_Applicative_5["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(g);
    };
};
var uncons = function (__dict_Monad_6) {
    return function (l) {
        var g = function (_382) {
            if (_382 instanceof Yield) {
                return Prelude.pure(__dict_Monad_6["__superclass_Prelude.Applicative_0"]())(Data_Maybe.Just.create(new Data_Tuple.Tuple(_382.value0, Data_Lazy.force(_382.value1))));
            };
            if (_382 instanceof Skip) {
                return uncons(__dict_Monad_6)(Data_Lazy.force(_382.value0));
            };
            if (_382 instanceof Done) {
                return Prelude.pure(__dict_Monad_6["__superclass_Prelude.Applicative_0"]())(Data_Maybe.Nothing.value);
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 180, column 3 - line 181, column 3: " + [ _382.constructor.name ]);
        };
        return Prelude[">>="](__dict_Monad_6["__superclass_Prelude.Bind_1"]())(runListT(l))(g);
    };
};

/**
 *  | Extract all but the first element of a list.
 */
var tail = function (__dict_Monad_7) {
    return function (l) {
        return Prelude["<$>"](((__dict_Monad_7["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude["<$>"](Data_Maybe.functorMaybe)(Data_Tuple.snd))(uncons(__dict_Monad_7)(l));
    };
};

/**
 *  | Prepend an element to a lazily-evaluated list.
 */
var prepend$prime = function (__dict_Applicative_8) {
    return function (h) {
        return function (t) {
            return ListT.create(Prelude.pure(__dict_Applicative_8)(new Yield(h, t)));
        };
    };
};

/**
 *  | Prepend an element to a list.
 */
var prepend = function (__dict_Applicative_9) {
    return function (h) {
        return function (t) {
            return prepend$prime(__dict_Applicative_9)(h)(Data_Lazy.defer(Prelude["const"](t)));
        };
    };
};

/**
 *  | The empty list.
 */
var nil = function (__dict_Applicative_10) {
    return ListT.create(Prelude.pure(__dict_Applicative_10)(Done.value));
};

/**
 *  | Create a list with one element.
 */
var singleton = function (__dict_Applicative_12) {
    return function (a) {
        return prepend(__dict_Applicative_12)(a)(nil(__dict_Applicative_12));
    };
};
var take = function (__dict_Applicative_13) {
    return function (_372) {
        return function (fa) {
            if (_372 === 0) {
                return nil(__dict_Applicative_13);
            };
            var f = function (_376) {
                if (_376 instanceof Yield) {
                    return new Yield(_376.value0, Prelude["<$>"](Data_Lazy.functorLazy)(take(__dict_Applicative_13)(Prelude["-"](Prelude.ringInt)(_372)(1)))(_376.value1));
                };
                if (_376 instanceof Skip) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(take(__dict_Applicative_13)(_372))(_376.value0));
                };
                if (_376 instanceof Done) {
                    return Done.value;
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 132, column 3 - line 133, column 3: " + [ _376.constructor.name ]);
            };
            return stepMap((__dict_Applicative_13["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(f)(fa);
        };
    };
};
var zipWith$prime = function (__dict_Monad_14) {
    return function (f) {
        var g = function (_387) {
            return function (_388) {
                if (_388 instanceof Data_Maybe.Nothing) {
                    return Prelude.pure(__dict_Monad_14["__superclass_Prelude.Applicative_0"]())(nil(__dict_Monad_14["__superclass_Prelude.Applicative_0"]()));
                };
                if (_387 instanceof Data_Maybe.Nothing) {
                    return Prelude.pure(__dict_Monad_14["__superclass_Prelude.Applicative_0"]())(nil(__dict_Monad_14["__superclass_Prelude.Applicative_0"]()));
                };
                if (_387 instanceof Data_Maybe.Just && _388 instanceof Data_Maybe.Just) {
                    return Prelude["<$>"](((__dict_Monad_14["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude.flip(prepend$prime(__dict_Monad_14["__superclass_Prelude.Applicative_0"]()))(Data_Lazy.defer(function (_369) {
                        return zipWith$prime(__dict_Monad_14)(f)(_387.value0.value1)(_388.value0.value1);
                    })))(f(_387.value0.value0)(_388.value0.value0));
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 222, column 3 - line 227, column 3: " + [ _387.constructor.name, _388.constructor.name ]);
            };
        };
        var loop = function (fa) {
            return function (fb) {
                return wrapEffect(((__dict_Monad_14["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude.bind(__dict_Monad_14["__superclass_Prelude.Bind_1"]())(uncons(__dict_Monad_14)(fa))(function (_64) {
                    return Prelude.bind(__dict_Monad_14["__superclass_Prelude.Bind_1"]())(uncons(__dict_Monad_14)(fb))(function (_63) {
                        return g(_64)(_63);
                    });
                }));
            };
        };
        return loop;
    };
};

/**
 *  | Zip the elements of two lists, combining elements at the same position from each list.
 */
var zipWith = function (__dict_Monad_15) {
    return function (f) {
        var g = function (a) {
            return function (b) {
                return Prelude.pure(__dict_Monad_15["__superclass_Prelude.Applicative_0"]())(f(a)(b));
            };
        };
        return zipWith$prime(__dict_Monad_15)(g);
    };
};
var mapMaybe = function (__dict_Functor_21) {
    return function (f) {
        var g = function (_381) {
            if (_381 instanceof Yield) {
                return Data_Maybe.fromMaybe(Skip.create)(Prelude["<$>"](Data_Maybe.functorMaybe)(Yield.create)(f(_381.value0)))(Prelude["<$>"](Data_Lazy.functorLazy)(mapMaybe(__dict_Functor_21)(f))(_381.value1));
            };
            if (_381 instanceof Skip) {
                return Skip.create(Prelude["<$>"](Data_Lazy.functorLazy)(mapMaybe(__dict_Functor_21)(f))(_381.value0));
            };
            if (_381 instanceof Done) {
                return Done.value;
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 168, column 3 - line 169, column 3: " + [ _381.constructor.name ]);
        };
        return stepMap(__dict_Functor_21)(g);
    };
};

/**
 *  | Generate an infinite list by iterating a function.
 */
var iterate = function (__dict_Monad_22) {
    return function (f) {
        return function (a) {
            var g = function (a_1) {
                return Prelude.pure(__dict_Monad_22["__superclass_Prelude.Applicative_0"]())(new Data_Maybe.Just(new Data_Tuple.Tuple(f(a_1), a_1)));
            };
            return unfold(__dict_Monad_22)(g)(a);
        };
    };
};

/**
 *  | Generate an infinite list by repeating a value.
 */
var repeat = function (__dict_Monad_23) {
    return iterate(__dict_Monad_23)(Prelude.id(Prelude.categoryFn));
};

/**
 *  | Extract the first element of a list.
 */
var head = function (__dict_Monad_24) {
    return function (l) {
        return Prelude["<$>"](((__dict_Monad_24["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude["<$>"](Data_Maybe.functorMaybe)(Data_Tuple.fst))(uncons(__dict_Monad_24)(l));
    };
};
var functorListT = function (__dict_Functor_25) {
    return new Prelude.Functor(function (f) {
        var g = function (_389) {
            if (_389 instanceof Yield) {
                return new Yield(f(_389.value0), Prelude["<$>"](Data_Lazy.functorLazy)(Prelude["<$>"](functorListT(__dict_Functor_25))(f))(_389.value1));
            };
            if (_389 instanceof Skip) {
                return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(Prelude["<$>"](functorListT(__dict_Functor_25))(f))(_389.value0));
            };
            if (_389 instanceof Done) {
                return Done.value;
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 245, column 5 - line 246, column 5: " + [ _389.constructor.name ]);
        };
        return stepMap(__dict_Functor_25)(g);
    });
};

/**
 *  | Lift a computation from the base functor.
 */
var fromEffect = function (__dict_Applicative_26) {
    return function (fa) {
        return ListT.create(Prelude["<$>"]((__dict_Applicative_26["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(function (_15) {
            return new Yield(_15, Data_Lazy.defer(function (_367) {
                return nil(__dict_Applicative_26);
            }));
        })(fa));
    };
};
var monadTransListT = new Control_Monad_Trans.MonadTrans(function (__dict_Monad_27) {
    return fromEffect(__dict_Monad_27["__superclass_Prelude.Applicative_0"]());
});

/**
 *  | Fold a list from the left, accumulating the result (effectfully) using the specified function.
 */
var foldl$prime = function (__dict_Monad_28) {
    return function (f) {
        var loop = function (b) {
            return function (l) {
                var g = function (_383) {
                    if (_383 instanceof Data_Maybe.Nothing) {
                        return Prelude.pure(__dict_Monad_28["__superclass_Prelude.Applicative_0"]())(b);
                    };
                    if (_383 instanceof Data_Maybe.Just) {
                        return Prelude[">>="](__dict_Monad_28["__superclass_Prelude.Bind_1"]())(f(b)(_383.value0.value0))(Prelude.flip(loop)(_383.value0.value1));
                    };
                    throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 197, column 5 - line 198, column 5: " + [ _383.constructor.name ]);
                };
                return Prelude[">>="](__dict_Monad_28["__superclass_Prelude.Bind_1"]())(uncons(__dict_Monad_28)(l))(g);
            };
        };
        return loop;
    };
};

/**
 *  | Fold a list from the left, accumulating the result using the specified function.
 */
var foldl = function (__dict_Monad_29) {
    return function (f) {
        var loop = function (b) {
            return function (l) {
                var g = function (_384) {
                    if (_384 instanceof Data_Maybe.Nothing) {
                        return Prelude.pure(__dict_Monad_29["__superclass_Prelude.Applicative_0"]())(b);
                    };
                    if (_384 instanceof Data_Maybe.Just) {
                        return loop(f(b)(_384.value0.value0))(_384.value0.value1);
                    };
                    throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 205, column 5 - line 206, column 5: " + [ _384.constructor.name ]);
                };
                return Prelude[">>="](__dict_Monad_29["__superclass_Prelude.Bind_1"]())(uncons(__dict_Monad_29)(l))(g);
            };
        };
        return loop;
    };
};
var filter = function (__dict_Functor_30) {
    return function (f) {
        var g = function (_380) {
            if (_380 instanceof Yield) {
                var s$prime = Prelude["<$>"](Data_Lazy.functorLazy)(filter(__dict_Functor_30)(f))(_380.value1);
                var _724 = f(_380.value0);
                if (_724) {
                    return new Yield(_380.value0, s$prime);
                };
                if (!_724) {
                    return new Skip(s$prime);
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 161, column 3 - line 162, column 3: " + [ _724.constructor.name ]);
            };
            if (_380 instanceof Skip) {
                var s$prime = Prelude["<$>"](Data_Lazy.functorLazy)(filter(__dict_Functor_30)(f))(_380.value0);
                return new Skip(s$prime);
            };
            if (_380 instanceof Done) {
                return Done.value;
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 161, column 3 - line 162, column 3: " + [ _380.constructor.name ]);
        };
        return stepMap(__dict_Functor_30)(g);
    };
};
var dropWhile = function (__dict_Applicative_31) {
    return function (f) {
        var g = function (_379) {
            if (_379 instanceof Yield) {
                var _729 = f(_379.value0);
                if (_729) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(dropWhile(__dict_Applicative_31)(f))(_379.value1));
                };
                if (!_729) {
                    return new Yield(_379.value0, _379.value1);
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 154, column 3 - line 155, column 3: " + [ _729.constructor.name ]);
            };
            if (_379 instanceof Skip) {
                return Skip.create(Prelude["<$>"](Data_Lazy.functorLazy)(dropWhile(__dict_Applicative_31)(f))(_379.value0));
            };
            if (_379 instanceof Done) {
                return Done.value;
            };
            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 154, column 3 - line 155, column 3: " + [ _379.constructor.name ]);
        };
        return stepMap((__dict_Applicative_31["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(g);
    };
};
var drop = function (__dict_Applicative_32) {
    return function (_373) {
        return function (fa) {
            if (_373 === 0) {
                return fa;
            };
            var f = function (_378) {
                if (_378 instanceof Yield) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(drop(__dict_Applicative_32)(Prelude["-"](Prelude.ringInt)(_373)(1)))(_378.value1));
                };
                if (_378 instanceof Skip) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(drop(__dict_Applicative_32)(_373))(_378.value0));
                };
                if (_378 instanceof Done) {
                    return Done.value;
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 147, column 3 - line 148, column 3: " + [ _378.constructor.name ]);
            };
            return stepMap((__dict_Applicative_32["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(f)(fa);
        };
    };
};

/**
 *  | Attach an element to the front of a list.
 */
var cons = function (__dict_Applicative_33) {
    return function (lh) {
        return function (t) {
            return ListT.create(Prelude.pure(__dict_Applicative_33)(new Yield(Data_Lazy.force(lh), t)));
        };
    };
};
var unfoldableListT = function (__dict_Monad_34) {
    return new Data_Unfoldable.Unfoldable(function (f) {
        return function (b) {
            var go = function (_390) {
                if (_390 instanceof Data_Maybe.Nothing) {
                    return nil(__dict_Monad_34["__superclass_Prelude.Applicative_0"]());
                };
                if (_390 instanceof Data_Maybe.Just) {
                    return cons(__dict_Monad_34["__superclass_Prelude.Applicative_0"]())(Prelude.pure(Data_Lazy.applicativeLazy)(_390.value0.value0))(Data_Lazy.defer(function (_370) {
                        return go(f(_390.value0.value1));
                    }));
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 249, column 1 - line 254, column 1: " + [ _390.constructor.name ]);
            };
            return go(f(b));
        };
    });
};
var semigroupListT = function (__dict_Applicative_36) {
    return new Prelude.Semigroup(concat(__dict_Applicative_36));
};
var concat = function (__dict_Applicative_35) {
    return function (x) {
        return function (y) {
            var f = function (_374) {
                if (_374 instanceof Yield) {
                    return new Yield(_374.value0, Prelude["<$>"](Data_Lazy.functorLazy)(function (_13) {
                        return Prelude["<>"](semigroupListT(__dict_Applicative_35))(_13)(y);
                    })(_374.value1));
                };
                if (_374 instanceof Skip) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(function (_14) {
                        return Prelude["<>"](semigroupListT(__dict_Applicative_35))(_14)(y);
                    })(_374.value0));
                };
                if (_374 instanceof Done) {
                    return new Skip(Data_Lazy.defer(Prelude["const"](y)));
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 91, column 3 - line 92, column 3: " + [ _374.constructor.name ]);
            };
            return stepMap((__dict_Applicative_35["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(f)(x);
        };
    };
};
var monoidListT = function (__dict_Applicative_16) {
    return new Data_Monoid.Monoid(function () {
        return semigroupListT(__dict_Applicative_16);
    }, nil(__dict_Applicative_16));
};

/**
 *  | Remove elements from a list which do not contain a value.
 */
var catMaybes = function (__dict_Functor_37) {
    return mapMaybe(__dict_Functor_37)(Prelude.id(Prelude.categoryFn));
};
var monadListT = function (__dict_Monad_18) {
    return new Prelude.Monad(function () {
        return applicativeListT(__dict_Monad_18);
    }, function () {
        return bindListT(__dict_Monad_18);
    });
};
var bindListT = function (__dict_Monad_38) {
    return new Prelude.Bind(function () {
        return applyListT(__dict_Monad_38);
    }, function (fa) {
        return function (f) {
            var g = function (_391) {
                if (_391 instanceof Yield) {
                    var h = function (s_1) {
                        return Prelude["<>"](semigroupListT(__dict_Monad_38["__superclass_Prelude.Applicative_0"]()))(f(_391.value0))(Prelude[">>="](bindListT(__dict_Monad_38))(s_1)(f));
                    };
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(h)(_391.value1));
                };
                if (_391 instanceof Skip) {
                    return new Skip(Prelude["<$>"](Data_Lazy.functorLazy)(function (_16) {
                        return Prelude[">>="](bindListT(__dict_Monad_38))(_16)(f);
                    })(_391.value0));
                };
                if (_391 instanceof Done) {
                    return Done.value;
                };
                throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/List/Trans.purs line 262, column 5 - line 265, column 5: " + [ _391.constructor.name ]);
            };
            return stepMap(((__dict_Monad_38["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(g)(fa);
        };
    });
};
var applyListT = function (__dict_Monad_39) {
    return new Prelude.Apply(function () {
        return functorListT(((__dict_Monad_39["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]());
    }, Prelude.ap(monadListT(__dict_Monad_39)));
};
var applicativeListT = function (__dict_Monad_40) {
    return new Prelude.Applicative(function () {
        return applyListT(__dict_Monad_40);
    }, singleton(__dict_Monad_40["__superclass_Prelude.Applicative_0"]()));
};
var monadEffListT = function (__dict_Monad_19) {
    return function (__dict_MonadEff_20) {
        return new Control_Monad_Eff_Class.MonadEff(function () {
            return monadListT(__dict_Monad_19);
        }, Prelude["<<<"](Prelude.semigroupoidFn)(Control_Monad_Trans.lift(monadTransListT)(__dict_Monad_19))(Control_Monad_Eff_Class.liftEff(__dict_MonadEff_20)));
    };
};
var altListT = function (__dict_Applicative_42) {
    return new Control_Alt.Alt(function () {
        return functorListT((__dict_Applicative_42["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]());
    }, concat(__dict_Applicative_42));
};
var plusListT = function (__dict_Monad_11) {
    return new Control_Plus.Plus(function () {
        return altListT(__dict_Monad_11["__superclass_Prelude.Applicative_0"]());
    }, nil(__dict_Monad_11["__superclass_Prelude.Applicative_0"]()));
};
var alternativeListT = function (__dict_Monad_41) {
    return new Control_Alternative.Alternative(function () {
        return plusListT(__dict_Monad_41);
    }, function () {
        return applicativeListT(__dict_Monad_41);
    });
};
var monadPlusListT = function (__dict_Monad_17) {
    return new Control_MonadPlus.MonadPlus(function () {
        return alternativeListT(__dict_Monad_17);
    }, function () {
        return monadListT(__dict_Monad_17);
    });
};
module.exports = {
    "zipWith'": zipWith$prime, 
    zipWith: zipWith, 
    wrapLazy: wrapLazy, 
    wrapEffect: wrapEffect, 
    unfold: unfold, 
    uncons: uncons, 
    takeWhile: takeWhile, 
    take: take, 
    tail: tail, 
    singleton: singleton, 
    scanl: scanl, 
    repeat: repeat, 
    "prepend'": prepend$prime, 
    prepend: prepend, 
    nil: nil, 
    mapMaybe: mapMaybe, 
    iterate: iterate, 
    head: head, 
    fromEffect: fromEffect, 
    "foldl'": foldl$prime, 
    foldl: foldl, 
    filter: filter, 
    dropWhile: dropWhile, 
    drop: drop, 
    cons: cons, 
    catMaybes: catMaybes, 
    semigroupListT: semigroupListT, 
    monoidListT: monoidListT, 
    functorListT: functorListT, 
    unfoldableListT: unfoldableListT, 
    applyListT: applyListT, 
    applicativeListT: applicativeListT, 
    bindListT: bindListT, 
    monadListT: monadListT, 
    monadTransListT: monadTransListT, 
    altListT: altListT, 
    plusListT: plusListT, 
    alternativeListT: alternativeListT, 
    monadPlusListT: monadPlusListT, 
    monadEffListT: monadEffListT
};
