// Control.Comonad.Traced.Trans/index.js

/**
 *  | This module defines the cowriter comonad transformer, `TracedT`.
 */
"use strict";
var Prelude = require("Prelude");
var Control_Extend = require("Control.Extend");
var Control_Comonad = require("Control.Comonad");
var Data_Monoid = require("Data.Monoid");
var Control_Comonad_Trans = require("Control.Comonad.Trans");
var Data_Tuple = require("Data.Tuple");

/**
 *  | The cowriter comonad transformer.
 *  |
 *  | This comonad transformer extends the context of a value in the base comonad so that the value
 *  | depends on a monoidal position of type `t`.
 *  |
 *  | The `ComonadTraced` type class describes the operations supported by this comonad.
 */
var TracedT = function (x) {
    return x;
};

/**
 *  | Unwrap a value in the `TracedT` comonad.
 */
var runTracedT = function (_308) {
    return _308;
};
var functorTracedT = function (__dict_Functor_0) {
    return new Prelude.Functor(function (f) {
        return function (_309) {
            return Prelude["<$>"](__dict_Functor_0)(function (g) {
                return function (t) {
                    return f(g(t));
                };
            })(_309);
        };
    });
};
var extendTracedT = function (__dict_Extend_1) {
    return function (__dict_Semigroup_2) {
        return new Control_Extend.Extend(function () {
            return functorTracedT(__dict_Extend_1["__superclass_Prelude.Functor_0"]());
        }, function (f) {
            return function (_310) {
                return Control_Extend["<<="](__dict_Extend_1)(function (w_1) {
                    return function (t) {
                        return f(Prelude["<$>"](__dict_Extend_1["__superclass_Prelude.Functor_0"]())(function (h) {
                            return function (t$prime) {
                                return h(Prelude["<>"](__dict_Semigroup_2)(t)(t$prime));
                            };
                        })(w_1));
                    };
                })(_310);
            };
        });
    };
};
var comonadTransTracedT = function (__dict_Monoid_3) {
    return new Control_Comonad_Trans.ComonadTrans(function (__dict_Comonad_4) {
        return function (_312) {
            return Prelude["<$>"]((__dict_Comonad_4["__superclass_Control.Extend.Extend_0"]())["__superclass_Prelude.Functor_0"]())(function (f) {
                return f(Data_Monoid.mempty(__dict_Monoid_3));
            })(_312);
        };
    });
};
var comonadTracedT = function (__dict_Comonad_5) {
    return function (__dict_Monoid_6) {
        return new Control_Comonad.Comonad(function () {
            return extendTracedT(__dict_Comonad_5["__superclass_Control.Extend.Extend_0"]())(__dict_Monoid_6["__superclass_Prelude.Semigroup_0"]());
        }, function (_311) {
            return Control_Comonad.extract(__dict_Comonad_5)(_311)(Data_Monoid.mempty(__dict_Monoid_6));
        });
    };
};
module.exports = {
    TracedT: TracedT, 
    runTracedT: runTracedT, 
    functorTracedT: functorTracedT, 
    extendTracedT: extendTracedT, 
    comonadTracedT: comonadTracedT, 
    comonadTransTracedT: comonadTransTracedT
};
