// Control.Comonad.Env.Trans/index.js

/**
 *  | This module defines the environment comonad transformer, `EnvT`.
 */
"use strict";
var Prelude = require("Prelude");
var Control_Extend = require("Control.Extend");
var Control_Comonad = require("Control.Comonad");
var Control_Comonad_Trans = require("Control.Comonad.Trans");
var Data_Tuple = require("Data.Tuple");

/**
 *  | The environment comonad transformer.
 *  |
 *  | This comonad transformer extends the context of a value in the base comonad with a _global environment_ of
 *  | type `e`.
 *  |
 *  | The `ComonadEnv` type class describes the operations supported by this comonad.
 */
var EnvT = function (x) {
    return x;
};

/**
 *  | Change the environment type in an `EnvT` context.
 */
var withEnvT = function (f) {
    return function (_294) {
        return EnvT(new Data_Tuple.Tuple(f(_294.value0), _294.value1));
    };
};

/**
 *  | Unwrap a value in the `EnvT` comonad.
 */
var runEnvT = function (_293) {
    return _293;
};

/**
 *  | Change the underlying comonad and data type in an `EnvT` context.
 */
var mapEnvT = function (f) {
    return function (_295) {
        return EnvT(new Data_Tuple.Tuple(_295.value0, f(_295.value1)));
    };
};
var functorEnvT = function (__dict_Functor_0) {
    return new Prelude.Functor(function (f) {
        return function (_296) {
            return EnvT(new Data_Tuple.Tuple(_296.value0, Prelude["<$>"](__dict_Functor_0)(f)(_296.value1)));
        };
    });
};
var extendEnvT = function (__dict_Extend_1) {
    return new Control_Extend.Extend(function () {
        return functorEnvT(__dict_Extend_1["__superclass_Prelude.Functor_0"]());
    }, function (f) {
        return function (_297) {
            return EnvT(new Data_Tuple.Tuple(_297.value0, Prelude["<$>"](__dict_Extend_1["__superclass_Prelude.Functor_0"]())(f)(Control_Extend["<<="](__dict_Extend_1)(Prelude[">>>"](Prelude.semigroupoidFn)(Data_Tuple.Tuple.create(_297.value0))(EnvT))(_297.value1))));
        };
    });
};
var comonadTransEnvT = new Control_Comonad_Trans.ComonadTrans(function (__dict_Comonad_2) {
    return function (_299) {
        return _299.value1;
    };
});
var comonadEnvT = function (__dict_Comonad_3) {
    return new Control_Comonad.Comonad(function () {
        return extendEnvT(__dict_Comonad_3["__superclass_Control.Extend.Extend_0"]());
    }, function (_298) {
        return Control_Comonad.extract(__dict_Comonad_3)(_298.value1);
    });
};
module.exports = {
    EnvT: EnvT, 
    mapEnvT: mapEnvT, 
    withEnvT: withEnvT, 
    runEnvT: runEnvT, 
    functorEnvT: functorEnvT, 
    extendEnvT: extendEnvT, 
    comonadEnvT: comonadEnvT, 
    comonadTransEnvT: comonadTransEnvT
};
