// Data.Bifunctor.Flip/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Bifunctor = require("Data.Bifunctor");
var Control_Biapply = require("Control.Biapply");
var Control_Biapplicative = require("Control.Biapplicative");

/**
 *  | Flips the order of the type arguments of a `Bifunctor`.
 */
var Flip = (function () {
    function Flip(value0) {
        this.value0 = value0;
    };
    Flip.create = function (value0) {
        return new Flip(value0);
    };
    return Flip;
})();

/**
 *  | Remove the `Flip` constructor.
 */
var runFlip = function (_119) {
    return _119.value0;
};
var flipFunctor = function (__dict_Bifunctor_0) {
    return new Prelude.Functor(function (f) {
        return Prelude["<<<"](Prelude.semigroupoidFn)(Flip.create)(Prelude["<<<"](Prelude.semigroupoidFn)(Data_Bifunctor.lmap(__dict_Bifunctor_0)(f))(runFlip));
    });
};
var flipBifunctor = function (__dict_Bifunctor_1) {
    return new Data_Bifunctor.Bifunctor(function (f) {
        return function (g) {
            return Prelude["<<<"](Prelude.semigroupoidFn)(Flip.create)(Prelude["<<<"](Prelude.semigroupoidFn)(Data_Bifunctor.bimap(__dict_Bifunctor_1)(g)(f))(runFlip));
        };
    });
};
var flipBiapply = function (__dict_Biapply_2) {
    return new Control_Biapply.Biapply(function () {
        return flipBifunctor(__dict_Biapply_2["__superclass_Data.Bifunctor.Bifunctor_0"]());
    }, function (_120) {
        return function (_121) {
            return new Flip(Control_Biapply["<<*>>"](__dict_Biapply_2)(_120.value0)(_121.value0));
        };
    });
};
var flipBiapplicative = function (__dict_Biapplicative_3) {
    return new Control_Biapplicative.Biapplicative(function () {
        return flipBiapply(__dict_Biapplicative_3["__superclass_Control.Biapply.Biapply_0"]());
    }, function (a) {
        return function (b) {
            return new Flip(Control_Biapplicative.bipure(__dict_Biapplicative_3)(b)(a));
        };
    });
};
module.exports = {
    Flip: Flip, 
    runFlip: runFlip, 
    flipBifunctor: flipBifunctor, 
    flipFunctor: flipFunctor, 
    flipBiapply: flipBiapply, 
    flipBiapplicative: flipBiapplicative
};
