// Control.Monad.Writer.Trans/index.js

/**
 *  | This module defines the writer monad transformer, `WriterT`.
 */
"use strict";
var Prelude = require("Prelude");
var Data_Tuple = require("Data.Tuple");
var Data_Monoid = require("Data.Monoid");
var Control_Alt = require("Control.Alt");
var Control_Plus = require("Control.Plus");
var Control_Monad_Rec_Class = require("Control.Monad.Rec.Class");
var Control_Monad_Trans = require("Control.Monad.Trans");
var Control_Monad_Eff_Class = require("Control.Monad.Eff.Class");
var Control_Alternative = require("Control.Alternative");
var Control_MonadPlus = require("Control.MonadPlus");
var Data_Either = require("Data.Either");

/**
 *  | The writer monad transformer.
 *  |
 *  | This monad transformer extends the base monad with a monoidal accumulator of
 *  | type `w`.
 *  |
 *  | The `MonadWriter` type class describes the operations supported by this monad.
 */
var WriterT = function (x) {
    return x;
};

/**
 *  | Run a computation in the `WriterT` monad.
 */
var runWriterT = function (_341) {
    return _341;
};
var monadTransWriterT = function (__dict_Monoid_4) {
    return new Control_Monad_Trans.MonadTrans(function (__dict_Monad_5) {
        return function (m) {
            return WriterT(Prelude.bind(__dict_Monad_5["__superclass_Prelude.Bind_1"]())(m)(function (_52) {
                return Prelude["return"](__dict_Monad_5["__superclass_Prelude.Applicative_0"]())(new Data_Tuple.Tuple(_52, Data_Monoid.mempty(__dict_Monoid_4)));
            }));
        };
    });
};

/**
 *  | Change the accumulator and base monad types in a `WriterT` monad action.
 */
var mapWriterT = function (f) {
    return function (m) {
        return WriterT(f(runWriterT(m)));
    };
};
var liftCatchWriter = function ($$catch) {
    return function (m) {
        return function (h) {
            return WriterT($$catch(runWriterT(m))(function (e) {
                return runWriterT(h(e));
            }));
        };
    };
};
var liftCallCCWriter = function (__dict_Monoid_13) {
    return function (callCC) {
        return function (f) {
            return WriterT(callCC(function (c) {
                return runWriterT(f(function (a) {
                    return WriterT(c(new Data_Tuple.Tuple(a, Data_Monoid.mempty(__dict_Monoid_13))));
                }));
            }));
        };
    };
};
var functorWriterT = function (__dict_Functor_14) {
    return new Prelude.Functor(function (f) {
        return mapWriterT(Prelude["<$>"](__dict_Functor_14)(function (_340) {
            return new Data_Tuple.Tuple(f(_340.value0), _340.value1);
        }));
    });
};

/**
 *  | Run a computation in the `WriterT` monad, discarding the result.
 */
var execWriterT = function (__dict_Apply_15) {
    return function (m) {
        return Prelude["<$>"](__dict_Apply_15["__superclass_Prelude.Functor_0"]())(Data_Tuple.snd)(runWriterT(m));
    };
};
var applyWriterT = function (__dict_Monoid_18) {
    return function (__dict_Apply_19) {
        return new Prelude.Apply(function () {
            return functorWriterT(__dict_Apply_19["__superclass_Prelude.Functor_0"]());
        }, function (f) {
            return function (v) {
                return WriterT((function () {
                    var k = function (_342) {
                        return function (_343) {
                            return new Data_Tuple.Tuple(_342.value0(_343.value0), Prelude["<>"](__dict_Monoid_18["__superclass_Prelude.Semigroup_0"]())(_342.value1)(_343.value1));
                        };
                    };
                    return Prelude["<*>"](__dict_Apply_19)(Prelude["<$>"](__dict_Apply_19["__superclass_Prelude.Functor_0"]())(k)(runWriterT(f)))(runWriterT(v));
                })());
            };
        });
    };
};
var bindWriterT = function (__dict_Monoid_16) {
    return function (__dict_Monad_17) {
        return new Prelude.Bind(function () {
            return applyWriterT(__dict_Monoid_16)((__dict_Monad_17["__superclass_Prelude.Applicative_0"]())["__superclass_Prelude.Apply_0"]());
        }, function (m) {
            return function (k) {
                return WriterT(Prelude.bind(__dict_Monad_17["__superclass_Prelude.Bind_1"]())(runWriterT(m))(function (_50) {
                    return Prelude.bind(__dict_Monad_17["__superclass_Prelude.Bind_1"]())(runWriterT(k(_50.value0)))(function (_49) {
                        return Prelude["return"](__dict_Monad_17["__superclass_Prelude.Applicative_0"]())(new Data_Tuple.Tuple(_49.value0, Prelude["<>"](__dict_Monoid_16["__superclass_Prelude.Semigroup_0"]())(_50.value1)(_49.value1)));
                    });
                }));
            };
        });
    };
};
var applicativeWriterT = function (__dict_Monoid_20) {
    return function (__dict_Applicative_21) {
        return new Prelude.Applicative(function () {
            return applyWriterT(__dict_Monoid_20)(__dict_Applicative_21["__superclass_Prelude.Apply_0"]());
        }, function (a) {
            return WriterT(Prelude.pure(__dict_Applicative_21)(new Data_Tuple.Tuple(a, Data_Monoid.mempty(__dict_Monoid_20))));
        });
    };
};
var monadWriterT = function (__dict_Monoid_2) {
    return function (__dict_Monad_3) {
        return new Prelude.Monad(function () {
            return applicativeWriterT(__dict_Monoid_2)(__dict_Monad_3["__superclass_Prelude.Applicative_0"]());
        }, function () {
            return bindWriterT(__dict_Monoid_2)(__dict_Monad_3);
        });
    };
};
var monadEffWriter = function (__dict_Monad_10) {
    return function (__dict_Monoid_11) {
        return function (__dict_MonadEff_12) {
            return new Control_Monad_Eff_Class.MonadEff(function () {
                return monadWriterT(__dict_Monoid_11)(__dict_Monad_10);
            }, Prelude["<<<"](Prelude.semigroupoidFn)(Control_Monad_Trans.lift(monadTransWriterT(__dict_Monoid_11))(__dict_Monad_10))(Control_Monad_Eff_Class.liftEff(__dict_MonadEff_12)));
        };
    };
};
var monadRecWriterT = function (__dict_Monoid_6) {
    return function (__dict_MonadRec_7) {
        return new Control_Monad_Rec_Class.MonadRec(function () {
            return monadWriterT(__dict_Monoid_6)(__dict_MonadRec_7["__superclass_Prelude.Monad_0"]());
        }, function (f) {
            return function (a) {
                var f$prime = function (_344) {
                    return Prelude.bind((__dict_MonadRec_7["__superclass_Prelude.Monad_0"]())["__superclass_Prelude.Bind_1"]())(runWriterT(f(_344.value0)))(function (_51) {
                        return Prelude["return"]((__dict_MonadRec_7["__superclass_Prelude.Monad_0"]())["__superclass_Prelude.Applicative_0"]())((function () {
                            if (_51.value0 instanceof Data_Either.Left) {
                                return new Data_Either.Left(new Data_Tuple.Tuple(_51.value0.value0, Prelude["<>"](__dict_Monoid_6["__superclass_Prelude.Semigroup_0"]())(_344.value1)(_51.value1)));
                            };
                            if (_51.value0 instanceof Data_Either.Right) {
                                return new Data_Either.Right(new Data_Tuple.Tuple(_51.value0.value0, Prelude["<>"](__dict_Monoid_6["__superclass_Prelude.Semigroup_0"]())(_344.value1)(_51.value1)));
                            };
                            throw new Error("Failed pattern match at bower_components/purescript-transformers/src/Control/Monad/Writer/Trans.purs line 68, column 5 - line 74, column 1: " + [ _51.value0.constructor.name ]);
                        })());
                    });
                };
                return WriterT(Control_Monad_Rec_Class.tailRecM(__dict_MonadRec_7)(f$prime)(new Data_Tuple.Tuple(a, Data_Monoid.mempty(__dict_Monoid_6))));
            };
        });
    };
};
var altWriterT = function (__dict_Monoid_24) {
    return function (__dict_Alt_25) {
        return new Control_Alt.Alt(function () {
            return functorWriterT(__dict_Alt_25["__superclass_Prelude.Functor_0"]());
        }, function (m) {
            return function (n) {
                return WriterT(Control_Alt["<|>"](__dict_Alt_25)(runWriterT(m))(runWriterT(n)));
            };
        });
    };
};
var plusWriterT = function (__dict_Monoid_0) {
    return function (__dict_Plus_1) {
        return new Control_Plus.Plus(function () {
            return altWriterT(__dict_Monoid_0)(__dict_Plus_1["__superclass_Control.Alt.Alt_0"]());
        }, Control_Plus.empty(__dict_Plus_1));
    };
};
var alternativeWriterT = function (__dict_Monoid_22) {
    return function (__dict_Alternative_23) {
        return new Control_Alternative.Alternative(function () {
            return plusWriterT(__dict_Monoid_22)(__dict_Alternative_23["__superclass_Control.Plus.Plus_1"]());
        }, function () {
            return applicativeWriterT(__dict_Monoid_22)(__dict_Alternative_23["__superclass_Prelude.Applicative_0"]());
        });
    };
};
var monadPlusWriterT = function (__dict_Monoid_8) {
    return function (__dict_MonadPlus_9) {
        return new Control_MonadPlus.MonadPlus(function () {
            return alternativeWriterT(__dict_Monoid_8)(__dict_MonadPlus_9["__superclass_Control.Alternative.Alternative_1"]());
        }, function () {
            return monadWriterT(__dict_Monoid_8)(__dict_MonadPlus_9["__superclass_Prelude.Monad_0"]());
        });
    };
};
module.exports = {
    WriterT: WriterT, 
    liftCallCCWriter: liftCallCCWriter, 
    liftCatchWriter: liftCatchWriter, 
    mapWriterT: mapWriterT, 
    execWriterT: execWriterT, 
    runWriterT: runWriterT, 
    functorWriterT: functorWriterT, 
    applyWriterT: applyWriterT, 
    applicativeWriterT: applicativeWriterT, 
    altWriterT: altWriterT, 
    plusWriterT: plusWriterT, 
    alternativeWriterT: alternativeWriterT, 
    bindWriterT: bindWriterT, 
    monadWriterT: monadWriterT, 
    monadRecWriterT: monadRecWriterT, 
    monadPlusWriterT: monadPlusWriterT, 
    monadTransWriterT: monadTransWriterT, 
    monadEffWriter: monadEffWriter
};
