// Test.Spec/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Trans = require("Control.Monad.Trans");
var Control_Monad_State_Class = require("Control.Monad.State.Class");
var Control_Monad_Aff = require("Control.Monad.Aff");
var Data_Either = require("Data.Either");
var Control_Monad_State_Trans = require("Control.Monad.State.Trans");
var Data_Tuple = require("Data.Tuple");
var Control_Monad_Eff_Exception = require("Control.Monad.Eff.Exception");
var Data_Monoid = require("Data.Monoid");
var Success = (function () {
    function Success() {

    };
    Success.value = new Success();
    return Success;
})();
var Failure = (function () {
    function Failure(value0) {
        this.value0 = value0;
    };
    Failure.create = function (value0) {
        return new Failure(value0);
    };
    return Failure;
})();
var Describe = (function () {
    function Describe(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Describe.create = function (value0) {
        return function (value1) {
            return new Describe(value0, value1);
        };
    };
    return Describe;
})();
var It = (function () {
    function It(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    It.create = function (value0) {
        return function (value1) {
            return new It(value0, value1);
        };
    };
    return It;
})();
var Pending = (function () {
    function Pending(value0) {
        this.value0 = value0;
    };
    Pending.create = function (value0) {
        return new Pending(value0);
    };
    return Pending;
})();
var showResult = new Prelude.Show(function (_142) {
    if (_142 instanceof Success) {
        return "Success";
    };
    if (_142 instanceof Failure) {
        return "Failure (Error ...)";
    };
    throw new Error("Failed pattern match at bower_components/purescript-spec/src/Test/Spec.purs line 32, column 1 - line 36, column 1: " + [ _142.constructor.name ]);
});
var showGroup = new Prelude.Show(function (_145) {
    if (_145 instanceof Describe) {
        return "Describe " + (Prelude.show(Prelude.showString)(_145.value0) + (" " + Prelude.show(Prelude.showArray(showGroup))(_145.value1)));
    };
    if (_145 instanceof It) {
        return "It " + (Prelude.show(Prelude.showString)(_145.value0) + (" " + Prelude.show(showResult)(_145.value1)));
    };
    if (_145 instanceof Pending) {
        return "Describe " + Prelude.show(Prelude.showString)(_145.value0);
    };
    throw new Error("Failed pattern match: " + [ _145.constructor.name ]);
});
var runCatch = function (name) {
    return function (tests) {
        var onSuccess = function (_148) {
            return Prelude["return"](Control_Monad_Aff.applicativeAff)(new It(name, Success.value));
        };
        var onError = function (e) {
            return Prelude["return"](Control_Monad_Aff.applicativeAff)(It.create(name)(new Failure(e)));
        };
        return Prelude.bind(Control_Monad_Aff.bindAff)(Control_Monad_Aff.attempt(tests))(function (_60) {
            return Data_Either.either(onError)(onSuccess)(_60);
        });
    };
};
var pending = function (name) {
    return Control_Monad_State_Class.modify(Control_Monad_State_Trans.monadStateT(Control_Monad_Aff.monadAff))(Control_Monad_State_Class.monadStateStateT(Control_Monad_Aff.monadAff))(function (p) {
        return Prelude["++"](Prelude.semigroupArray)(p)([ new Pending(name) ]);
    });
};
var it = function (description) {
    return function (tests) {
        return Prelude.bind(Control_Monad_State_Trans.bindStateT(Control_Monad_Aff.monadAff))(Control_Monad_Trans.lift(Control_Monad_State_Trans.monadTransStateT)(Control_Monad_Aff.monadAff)(runCatch(description)(tests)))(function (_61) {
            return Prelude.bind(Control_Monad_State_Trans.bindStateT(Control_Monad_Aff.monadAff))(Control_Monad_State_Class.modify(Control_Monad_State_Trans.monadStateT(Control_Monad_Aff.monadAff))(Control_Monad_State_Class.monadStateStateT(Control_Monad_Aff.monadAff))(function (p) {
                return Prelude["++"](Prelude.semigroupArray)(p)([ _61 ]);
            }))(function () {
                return Prelude["return"](Control_Monad_State_Trans.applicativeStateT(Control_Monad_Aff.monadAff))(Prelude.unit);
            });
        });
    };
};
var eqResult = new Prelude.Eq(function (_143) {
    return function (_144) {
        if (_143 instanceof Success && _144 instanceof Success) {
            return true;
        };
        if (_143 instanceof Failure && _144 instanceof Failure) {
            return true;
        };
        return false;
    };
});
var eqGroup = new Prelude.Eq(function (_146) {
    return function (_147) {
        if (_146 instanceof Describe && _147 instanceof Describe) {
            return Prelude["&&"](Prelude.booleanAlgebraBoolean)(Prelude["=="](Prelude.eqString)(_146.value0)(_147.value0))(Prelude["=="](Prelude.eqArray(eqGroup))(_146.value1)(_147.value1));
        };
        if (_146 instanceof It && _147 instanceof It) {
            return Prelude["&&"](Prelude.booleanAlgebraBoolean)(Prelude["=="](Prelude.eqString)(_146.value0)(_147.value0))(Prelude["=="](eqResult)(_146.value1)(_147.value1));
        };
        if (_146 instanceof Pending && _147 instanceof Pending) {
            return Prelude["=="](Prelude.eqString)(_146.value0)(_147.value0);
        };
        return false;
    };
});
var collect = function (r) {
    return Prelude.bind(Control_Monad_Aff.bindAff)(Control_Monad_State_Trans.runStateT(r)([  ]))(function (_62) {
        return Prelude["return"](Control_Monad_Aff.applicativeAff)(Data_Tuple.snd(_62));
    });
};
var describe = function (name) {
    return function (its) {
        return Prelude.bind(Control_Monad_State_Trans.bindStateT(Control_Monad_Aff.monadAff))(Control_Monad_Trans.lift(Control_Monad_State_Trans.monadTransStateT)(Control_Monad_Aff.monadAff)(collect(its)))(function (_59) {
            return Prelude.bind(Control_Monad_State_Trans.bindStateT(Control_Monad_Aff.monadAff))(Control_Monad_State_Class.modify(Control_Monad_State_Trans.monadStateT(Control_Monad_Aff.monadAff))(Control_Monad_State_Class.monadStateStateT(Control_Monad_Aff.monadAff))(function (r) {
                return Prelude["++"](Prelude.semigroupArray)(r)([ new Describe(name, _59) ]);
            }))(function () {
                return Prelude["return"](Control_Monad_State_Trans.applicativeStateT(Control_Monad_Aff.monadAff))(Prelude.unit);
            });
        });
    };
};
module.exports = {
    Describe: Describe, 
    It: It, 
    Pending: Pending, 
    Success: Success, 
    Failure: Failure, 
    collect: collect, 
    it: it, 
    pending: pending, 
    describe: describe, 
    showResult: showResult, 
    eqResult: eqResult, 
    showGroup: showGroup, 
    eqGroup: eqGroup
};
